src/guest_memory.rs

```diff
@@ -302,6 +306,96 @@ pub trait GuestMemoryMut {
     }
 }
 
+/// GuestMemorySnapshottable provides a way to retrieve a GuestMemory object.
+/// The vm-memory crate already provides trivial implementation for
+/// instances of GuestMemory (more precisely, for references to them
+/// or reference-counted GuestMemory objects), but the trait can also
+/// be implemented by a GuestMemoryMut object in order to obtain a
+/// consistent snapshot of the memory map.
+///
+/// In order to support generic mutable memory maps, devices (or other things
+/// that access memory) should store the memory as a GuestMemorySnapshottable<M>.
+/// In the following example, dummy structs that implement the GuestMemoryBackend trait
+/// describe how the VMM is storing the memory map.  We could also declare VirtioDevice
+/// as `<M: GuestMemory, S: GuestMemorySnapshottable<M>`, but this would lead
+/// to repetitive types such as `VirtioDevice<GuestMemoryMmap, &GuestMemoryMmap>`.
+///
+/// ```
+/// # #[cfg(feature = "backend-mmap")]
+/// # use vm_memory::GuestMemoryMmap;
+/// # use vm_memory::{GuestAddress, GuestMemory, GuestMemorySnapshottable};
+///
+/// # // Perhaps this could be added to vm-memory itself?
+/// trait GuestMemoryBackend<'a> {
+///   type Snapshot: GuestMemory;
+///   type Memory: 'a + GuestMemorySnapshottable<Self::Snapshot>;
+/// }
+///
+/// pub struct VirtioDevice<'a, MB: GuestMemoryBackend<'a>> {
+///     mem: Option<MB::Memory>,
+/// }
+///
+/// impl<'a, MB: GuestMemoryBackend<'a>> VirtioDevice<'a, MB> {
+///     fn new() -> Self {
+///         VirtioDevice { mem: None }
+///     }
+///     fn activate(&mut self, mem: MB::Memory) {
+///         self.mem = Some(mem)
+///     }
+/// }
+///
+/// # #[cfg(feature = "backend-mmap")]
+/// # fn get_mmap() -> GuestMemoryMmap {
+/// #     GuestMemoryMmap::new(&[(GuestAddress(0),0)]).unwrap()
+/// # }
+///
+/// # #[cfg(feature = "backend-mmap")]
+/// # fn test_1() {
+/// // Using `VirtioDevice` with an immutable GuestMemoryMmap:
+/// struct ImmutableMmap;
+/// impl<'a> GuestMemoryBackend<'a> for ImmutableMmap {
+///     type Snapshot = GuestMemoryMmap;
+///     type Memory = &'a GuestMemoryMmap;
+/// }
+/// let mut for_immutable_mmap: VirtioDevice<ImmutableMmap> =
+///     VirtioDevice::new();
+/// let mmap = get_mmap();
+/// for_immutable_mmap.activate(&mmap);
+/// # }
+/// ```
+
+pub trait GuestMemorySnapshottable<M: GuestMemory> {
+    /// The type of the object returned by `snapshot()`.
+    type S: Deref<Target = M>;
+
+    /// Return a consistent snapshot of the memory map.
+    fn snapshot(&self) -> Self::S;
```

Comment:
The biggest question in my mind right now is whether this method needs to be a first class citizen of the `GuestMemory` interface. I can think of two potential reasons:
- There are use cases where the region set needs to remain immutable over the course of multiple `GuestMemory` method invocations. Like I mentioned in the `GuestMemoryMut` comment, I'm really wondering why that would be necessary, so any examples are greatly appreciated.
- The cost of the `snapshot` operation is significant enough that we want/need to amortize it over multiple inner `GuestMemory` method calls whenever possible.

If either of the above is true, I think this needs to be a part of the main `GuestMemory` interface, because components such as device building blocks that are written without assuming a particular guest memory implementation need to use a uniform interface (I think one of Gerry's comments was also pointing this out). Otherwise, it can be abstracted within a wrapper implementation (such as `GuestMemoryAtomic`) of the regular `GuestMemory` interface.

Reply:
> The cost of the snapshot operation is significant enough that we want/need to amortize it over multiple inner GuestMemory method calls whenever possible.

It's not really the cost of snapshot(), it's the cost of find_region(). And if you want to cache the result of find_region(), you cannot use the do_in_region() approach and you need a guard. snapshot() provides the guard.

> If either of the above is true, I think this needs to be a part of the main GuestMemory interface

This method is a part of the main interface, though it's a bit roundabout---more precisely,  thanks to the generic implementation of `GuestMemorySnapshottable` it is a part of `&'a GuestMemory` and `Arc<GuestMemory>`, for both of which it can simply return `self.